### PR TITLE
Add support for ESP32-C3 in esp-hal

### PR DESCRIPTION
**Pull request by misery**
# What does this implement/fix? 
Fixes some compile errors with esp32-c3 build.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
  
# Test Environment

- [ ] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
esphome:
  name: esp_c3
  platform: ESP32
  board: nodemcu-32s
  platformio_options:
    board_build.mcu: esp32c3
    platform: https://github.com/Jason2866/platform-espressif32.git#feature/arduino-c3
    platform_packages: framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#master
```

# Explain your changes
https://github.com/esphome/feature-requests/issues/1217

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by misery** [comment]
Any idea with that linker error?

```
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: .pioenvs/esp_demo_c3/libFrameworkArduino.a(wiring_pulse.c.o): in function `pulseIn':
/home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:38: undefined reference to `xthal_get_ccount'
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:40: undefined reference to `xthal_get_ccount'
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:45: undefined reference to `xthal_get_ccount'
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:41: undefined reference to `xthal_get_ccount'
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:42: undefined reference to `xthal_get_ccount'
/home/andre/.platformio/packages/toolchain-riscv32/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: .pioenvs/esp_demo_c3/libFrameworkArduino.a(wiring_pulse.c.o):/home/andre/.platformio/packages/framework-arduinoespressif32@src-cba3def1496a47e6af73c0b73bd2e13c/cores/esp32/wiring_pulse.c:43: more undefined references to `xthal_get_ccount' follow
```

File: esphome/core/esphal.cpp [hunk 1]
```diff
@@ -29,9 +29,15 @@ GPIOPin::GPIOPin(uint8_t pin, uint8_t mode, bool inverted)
       gpio_mask_(pin < 16 ? (1UL << pin) : 1)
 #endif
 #ifdef ARDUINO_ARCH_ESP32
+#ifdef CONFIG_IDF_TARGET_ESP32C3
+          gpio_set_(pin < 32 ? &GPIO.out_w1ts.val : &GPIO.out_w1tc.val),
```
Comment (agners) [inline esphome/core/esphal.cpp:33]: I think this should just be:
```suggestion
          gpio_set_(&GPIO.out_w1ts.val),
```

**Review by agners — CHANGES_REQUESTED** [review]
I also get this:

```
src/esphome/components/api/user_services.cpp: In function 'T esphome::api::get_execute_arg_value(const esphome::api::ExecuteServiceArgument&) [with T = std::vector<int>]':
src/esphome/components/api/user_services.cpp:19:14: error: could not convert 'arg.esphome::api::ExecuteServiceArgument::int_array' from 'vector<long int>' to 'vector<int>'
   return arg.int_array;
          ~~~~^~~~~~~~~
```
Which should be easily fixable with this change:
```
--- a/esphome/components/api/user_services.cpp
+++ b/esphome/components/api/user_services.cpp
@@ -15,7 +15,7 @@ template<> std::string get_execute_arg_value<std::string>(const ExecuteServiceAr
 template<> std::vector<bool> get_execute_arg_value<std::vector<bool>>(const ExecuteServiceArgument &arg) {
   return arg.bool_array;
 }
-template<> std::vector<int> get_execute_arg_value<std::vector<int>>(const ExecuteServiceArgument &arg) {
+template<> std::vector<int32_t> get_execute_arg_value<std::vector<int32_t>>(const ExecuteServiceArgument &arg) {
   return arg.int_array;
 }
 template<> std::vector<float> get_execute_arg_value<std::vector<float>>(const ExecuteServiceArgument &arg) {
```

I don't see the linking errors you mentioned though.

**Comment by oxan** [comment]
It seems to me this one is superseded by #2035 and is no longer necessary.

**Pull request closed without merging**